content/en/platform/corda/5.0/developing-applications/getting-started/utxo-advanced-ledger-extensions/advanced-design.md

```diff
@@ -0,0 +1,413 @@
+  
+   @Nullable
+   
+   private final StateRef id;
+  
+   public ExampleIdentifiableState(@Nullable final StateRef id) {
+    this.id = id;
+  }
+  
+   @Nullable
+  
+   public StateRef getId() {
+    return id;
+  }
+  
+   @NotNull
+   
+    public List<PublicKey> getParticipants() {
+    return List.of(...);
+  }
+}
+```
+
+## Designing Identifiable Commands
+
+Identifiable commands support creating, updating and deleting identifiable states.
+The `IdentifiableContractCreateCommand` supports creating new identifiable states and verifies the identifiable state(s) creation, at least one identifiable state must be created.
+
+```kotlin
+public final class Create extends IdentifiableContractCreateCommand<ExampleIdentifiableState> {
+  
+   @NotNull
+   
+   public Class<ExampleIdentifiableState> getContractStateType() {
+   return ExampleIdentifiableState.class;
+  }
+  
+  @Override
+  
+  protected void onVerify(@NotNull final UtxoLedgerTransaction transaction) {
+  
+    // Verify additional Create constraints
+    
+  }
+}
+```
+
+The `IdentifiableContractUpdateCommand` updates existing identifiable states and will verify the following constraints:
+
+* On identifiable state(s) updating, at least one identifiable state must be consumed.
+* On identifiable state(s) updating, at least one identifiable state must be created.
+* On identifiable state(s) updating, each created identifiable state's identifier must match one consumed identifiable state's state ref or identifier, exclusively.
+
+```kotlin
+public final class Update extends IdentifiableContractUpdateCommand<ExampleIdentifiableState> {
+  
+  @NotNull
+  
+   public Class<ExampleIdentifiableState> getContractStateType() {
+   return ExampleIdentifiableState.class;
+  }
+  
+  @Override
+  
+  protected void onVerify(@NotNull final UtxoLedgerTransaction transaction) {
+  
+  // Verify additional Update constraints
+  
+  }
+}
+```
+
+The `IdentifiableContractDeleteCommand` deletes existing identifiable states and verifies the identifiable state(s) deletion, at least one identifiable state must be consumed.
+
+```kotlin
+public final class Delete extends IdentifiableContractDeleteCommand<ExampleIdentifiableState> {
+  
+   @NotNull
+   
+   public Class<ExampleIdentifiableState> getContractStateType() {
+    return ExampleIdentifiableState.class;
+  }
+  
+   @Override
+   
+    protected void onVerify(@NotNull final UtxoLedgerTransaction transaction) {
+    
+    // Verify additional Delete constraints
+    
+  }
+}
+```
+
+## Designing an Identifiable Contract
+
+An identifiable contract can be implemented by extending the `IdentifiableContract` class, for example:
+
+```kotlin
+public final class ExampleIdentifiableContract extends IdentifiableContract {
+  
+   @Override
+   
+   public List<Class<? extends IdentifiableContractCommand<?>>> getPermittedCommandTypes() {
+   return List.of(Create.class, Update.class, Delete.class);
+   
+  }
+}
+```
```

Comment:
```suggestion
```kotlin
public final class ExampleIdentifiableContract extends IdentifiableContract {
  @Override
  public List<Class<? extends IdentifiableContractCommand<?>>> getPermittedCommandTypes() {
    return List.of(Create.class, Update.class, Delete.class);
  }
}
```
```